File: front-end/Mobile/src/models/attendance/AttendanceRecord.js
import BaseModel from '../BaseModel';

export default class AttendanceRecord extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.attendanceRecordId = data.attendance_record_id || null;
    this.classSessionId = data.class_session_id || null;
    this.academicActorId = data.academic_actor_id || null;
    this.attendanceStatus = data.attendance_status || 'Absent';
    this.capturedAt = data.captured_at || null;
    this.captureMethod = data.capture_method || 'MANUAL';
    this.matchScore = data.match_score || null;
  }

  get isPresent() {
    return this.attendanceStatus === 'Present';
  }

  get isAbsent() {
    return this.attendanceStatus === 'Absent';
  }

  get isLate() {
    return this.attendanceStatus === 'Late';
  }

  get isJustified() {
    return this.attendanceStatus === 'Justified';
  }

  get statusColor() {
    const colors = {
      Present: '#2da351',
      Absent: '#ff0000',
      Late: '#E65100',
      Justified: '#2196F3',
    };
    return colors[this.attendanceStatus] || '#999999';
  }

  static fromApi(data) {
    if (!data) return null;
    return new AttendanceRecord(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      attendance_record_id: this.attendanceRecordId,
      class_session_id: this.classSessionId,
      academic_actor_id: this.academicActorId,
      attendance_status: this.attendanceStatus,
      captured_at: this.capturedAt,
      capture_method: this.captureMethod,
      match_score: this.matchScore,
    };
  }
}

File: back-end/05-ms-attendance/src/main/java/com/faceattend_edu/attendance_service/domain/model/AttendanceRecord.java
package com.faceattend_edu.attendance_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Getter @Setter @AllArgsConstructor @NoArgsConstructor
public class AttendanceRecord {
    private Long attendanceRecordId;
    private Long classSessionId;
    private Long academicActorId;
    private String attendanceStatus;
    private String captureMethod;
    private Instant capturedAt;
    private BigDecimal matchScore;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    private static final Set<String> ALLOWED_STATUS = Set.of("Present","Absent","Late","Justified");
    private static final Set<String> ALLOWED_METHOD = Set.of("FACIAL","MANUAL","IOT","IMPORT");

    public void validate() {
        if (classSessionId == null) throw new IllegalArgumentException("classSessionId is required");
        if (academicActorId == null) throw new IllegalArgumentException("academicActorId is required");
        if (attendanceStatus == null || attendanceStatus.isBlank()) throw new IllegalArgumentException("attendanceStatus is required");
        if (!ALLOWED_STATUS.contains(attendanceStatus)) throw new IllegalArgumentException("attendanceStatus must be one of " + ALLOWED_STATUS);
        if (captureMethod == null || captureMethod.isBlank()) throw new IllegalArgumentException("captureMethod is required");
        if (!ALLOWED_METHOD.contains(captureMethod)) throw new IllegalArgumentException("captureMethod must be one of " + ALLOWED_METHOD);
        if (matchScore != null && (matchScore.compareTo(BigDecimal.ZERO) < 0 || matchScore.compareTo(BigDecimal.ONE) > 0))
            throw new IllegalArgumentException("matchScore must be between 0 and 1");
    }

    public void touchCreated() {
        if (createdAt == null) createdAt = Instant.now();
        if (capturedAt == null) capturedAt = Instant.now();
        if (rowVersion == 0) rowVersion = 1L;
    }
    public void touchUpdated() { updatedAt = Instant.now(); rowVersion++; }

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AttendanceRecord)) return false;
        AttendanceRecord that = (AttendanceRecord) o;
        return Objects.equals(attendanceRecordId, that.attendanceRecordId);
    }
    @Override public int hashCode() { return Objects.hash(attendanceRecordId); }
}
